# Introduction to Python and Machine Learning 
## By Chukwuma Anthony Nwachukwu
## anthonynwachukwu19@gmail.com

In [64]:
#importing libraries
import pandas as pd
import numpy as np

In [ ]:
#importing dataset
data=pd.read_csv('C:\Users\TonyC\Desktop\Intern Projects\Introduction to Python for Machine learning')

In [26]:
data.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

In [10]:
data.tail()

record_id  utility_id_ferc1  report_year  \
29518  f1_fuel_2018_12_12_0_13                12         2018   
29519   f1_fuel_2018_12_12_1_1                12         2018   
29520  f1_fuel_2018_12_12_1_10                12         2018   
29521  f1_fuel_2018_12_12_1_13                12         2018   
29522  f1_fuel_2018_12_12_1_14                12         2018   

           plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
29518    neil simpson ct #1                 gas       mcf          18799.0   
29519  cheyenne prairie 58%                 gas       mcf         806730.0   
29520     lange ct facility                 gas       mcf         104554.0   
29521       wygen 3 bhp 52%                coal       ton         315945.0   
29522       wygen 3 bhp 52%                 gas       mcf          17853.0   

       fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
29518                1.059                       4.78   
29519                1.050                       3.65   
29520                1.060                       4.77   
29521               16.108                       3.06   
29522                1.059                       0.00   

       fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
29518                          4.78                 9.03  
29519                          3.65                 6.95  
29520                          4.77                 8.99  
29521                         14.76                 1.11  
29522                          0.00                11.68

In [40]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29523 entries, 0 to 29522
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   record_id                     29523 non-null  object 
 1   utility_id_ferc1              29523 non-null  int64  
 2   report_year                   29523 non-null  int64  
 3   plant_name_ferc1              29523 non-null  object 
 4   fuel_type_code_pudl           29523 non-null  object 
 5   fuel_unit                     29343 non-null  object 
 6   fuel_qty_burned               29523 non-null  float64
 7   fuel_mmbtu_per_unit           29523 non-null  float64
 8   fuel_cost_per_unit_burned     29523 non-null  float64
 9   fuel_cost_per_unit_delivered  29523 non-null  float64
 10  fuel_cost_per_mmbtu           29523 non-null  float64
dtypes: float64(5), int64(2), object(4)
memory usage: 2.5+ MB


__Q4__

In [14]:
# Calculates mean,count,std,min,max,25,50,75 percentile.
round(data.describe(),2)

utility_id_ferc1  report_year  fuel_qty_burned  fuel_mmbtu_per_unit  \
count          29523.00     29523.00     2.952300e+04             29523.00   
mean             118.60      2005.81     2.622119e+06                 8.49   
std               74.18         7.03     9.118004e+06                10.60   
min                1.00      1994.00     1.000000e+00                 0.00   
25%               55.00      2000.00     1.381700e+04                 1.02   
50%              122.00      2006.00     2.533220e+05                 5.76   
75%              176.00      2012.00     1.424034e+06                17.01   
max              514.00      2018.00     5.558942e+08               341.26   

       fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  \
count                   29523.00                      29523.00   
mean                      208.65                        917.57   
std                      2854.49                      68775.93   
min                      -276.08                       -874.94   
25%                         5.21                          3.78   
50%                        26.00                         17.37   
75%                        47.11                         42.14   
max                    139358.00                    7964521.00   

       fuel_cost_per_mmbtu  
count             29523.00  
mean                 19.30  
std                2091.54  
min                 -41.50  
25%                   1.94  
50%                   4.13  
75%                   7.74  
max              359278.00

__Standard deviation and 75th percentile for Fuel_mmbtu_per_unit is 10.60 and 17.01 respectively. __

In [28]:
# checks the number of missing values from the dataset
data.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

__Q6.__

In [31]:
#Column with missing values
Column=data.columns[data.isnull().any()].tolist() 

# Total number of missing values
total=data.isnull().sum().sum() 

#percent of missing row/total number of rows in 3 decimal points
percent=round(data.isnull().sum().sum()/data.shape[0]*100,3) 

print('Feature:',Column,', ''Total:',total,', ''Percent:',percent)

Feature: ['fuel_unit'] , Total: 180 , Percent: 0.61


__Q7.__

In [65]:
# checks the feature under fuel_unit with the highest value
data.groupby("fuel_unit")["fuel_unit"].count

#Fill the missing data from the fuel_unit with the mode's value
data[["fuel_unit"]]=data[['fuel_unit']].fillna(value='mcf')

__Feature with missing values is from a Categorical header, The missing data can be filled/resolved using the Mode imputation.__

__Q3.__

In [51]:
#checks the mean of a feature(fuel_type_code_pudl)
round(data.groupby("fuel_type_code_pudl")["fuel_cost_per_unit_burned"].mean(),2).sort_values(ascending=True)

fuel_type_code_pudl
gas          13.66
other        18.25
waste        19.52
coal         67.42
oil         168.88
nuclear    4955.16
Name: fuel_cost_per_unit_burned, dtype: float64

__Gas has the lowest mean value or the least average fuel cost per unit burned.__ 

__Q5.__ 

In [38]:
#Calculates the skewness of features
round(data.skew(),2)

utility_id_ferc1                  0.61
report_year                       0.01
fuel_qty_burned                  15.85
fuel_mmbtu_per_unit               4.14
fuel_cost_per_unit_burned        19.79
fuel_cost_per_unit_delivered    105.01
fuel_cost_per_mmbtu             171.68
dtype: float64

In [39]:
# Calculates the Kurtiosis of features
round(data.kurtosis(),2)

utility_id_ferc1                    1.09
report_year                        -1.15
fuel_qty_burned                   651.37
fuel_mmbtu_per_unit                55.60
fuel_cost_per_unit_burned         485.26
fuel_cost_per_unit_delivered    11765.05
fuel_cost_per_mmbtu             29489.13
dtype: float64

__Skewness and Kurtiosis for fuel quantity burned are 15.85 and 651.37 respectively._

__Q8.__ 

In [42]:
#calculates correlation of features
data.corr()['fuel_cost_per_unit_burned'].sort_values(ascending=True)

utility_id_ferc1               -0.037863
fuel_qty_burned                -0.018535
fuel_mmbtu_per_unit            -0.010034
fuel_cost_per_mmbtu            -0.000437
fuel_cost_per_unit_delivered    0.011007
report_year                     0.013599
fuel_cost_per_unit_burned       1.000000
Name: fuel_cost_per_unit_burned, dtype: float64

__Feature with Second and Third lowest Correlation values are fuel_qty_burned and fuel_mmbtu_per_unit.__

__Q10__

In [41]:
#calculates mean for fuel_cost_per_unit_delivered by report_year
data.groupby("report_year")['fuel_cost_per_unit_delivered'].mean().sort_values(ascending=False)

report_year
1997    11140.197239
1996     9196.705948
2000      985.362877
2009      652.694163
2018      499.269966
2015      326.535511
1998      287.154420
2014      192.737183
2013      172.307591
2004      139.524275
2016      103.901761
2010       91.862105
1994       63.636060
2012       60.994502
2001       60.050396
2011       59.774667
2008       58.588197
2003       55.663493
2002       47.594361
2017       46.196861
2007       43.325023
2005       41.438184
2006       38.657484
1995       32.735269
1999       25.551627
Name: fuel_cost_per_unit_delivered, dtype: float64

__The report year with highest mean value for fuel unit per unit delivered is 1997.__ 

__Q9.__ 

In [47]:
# Calculates the sum of the fuel unit burned for fuel type coal for 1998
a=data.loc[(data['report_year']==1998) & (data['fuel_type_code_pudl']=='coal')]['fuel_cost_per_unit_burned'].sum()

# Calculates the sum of the fuel unit burned for fuel type coal for 1994
b=data.loc[(data['report_year']==1994) & (data['fuel_type_code_pudl']=='coal')]['fuel_cost_per_unit_burned'].sum()

# Calculates and prints percentage change 
print("Percentage Change =",round(((a-b)/a)*100),'%')

Percentage Change = -26 %


__Answer: None of the above. Percentage change= 26.0%__

__Q1.__

In [58]:
A = [1,2,3,4,5,6]
B = [13, 21, 34]
A.extend(B)
A

[1, 2, 3, 4, 5, 6, 13, 21, 34]

__Answer: A.expend(B)__

__Q2.__

In [60]:
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [61]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

__np.identity(3) and np.eye(3) creates an identity matrics__

# End of Notebook.